crates/ntx-builder/src/state/mod.rs

```diff
@@ -0,0 +1,278 @@
+            return;
+        }
+
+        let mut tx_impact = Impact::default();
+        if let Some(update) = account_delta.and_then(NetworkAccountUpdate::from_protocol) {
+            tx_impact.account_delta = Some(update.prefix());
+            self.account_or_default(update.prefix()).add_delta(update);
+        }
+        for note in network_notes {
+            tx_impact.notes.insert(note.nullifier());
+            self.nullifier_idx.insert(note.nullifier(), note.account_prefix());
+            self.account_or_default(note.account_prefix()).add_note(note);
+        }
+        for nullifier in nullifiers {
+            // Ignore nullifiers that aren't network note nullifiers.
+            let Some(account) = self.nullifier_idx.get(&nullifier) else {
+                continue;
+            };
+            tx_impact.nullifiers.insert(nullifier);
+            // We don't use the entry wrapper here because the account must already exist.
+            self.accounts
+                .get_mut(account)
+                .expect("nullifier account must exist")
+                .add_nullifier(nullifier);
+        }
+
+        if !tx_impact.is_empty() {
+            self.inflight_txs.insert(id, tx_impact);
+        }
+    }
+
+    /// Grants mutable access to the given account state, creating a default entry if none exists.
+    ///
+    /// This is effectively a thin wrapper around the entry API, but this also tracks new
+    /// accounts to the queue.
+    ///
+    /// This _must_ be the only way new accounts are added as otherwise they won't be queued.
+    fn account_or_default(&mut self, prefix: NetworkAccountPrefix) -> &mut AccountState {
+        match self.accounts.entry(prefix) {
+            Entry::Occupied(occupied_entry) => occupied_entry.into_mut(),
+            Entry::Vacant(vacant_entry) => {
+                self.queue.push_back(prefix);
+                vacant_entry.insert(AccountState::default())
+            },
+        }
+    }
+
+    /// Handles [`MempoolEvent::BlockCommitted`] events.
+    fn commit_transaction(&mut self, tx: TransactionId) {
+        let Impact { account_delta, notes: _, nullifiers } =
+            self.inflight_txs.remove(&tx).unwrap_or_default();
+
+        if let Some(prefix) = account_delta {
+            if self.accounts.get_mut(&prefix).unwrap().commit_delta().is_empty() {
+                self.remove_account(prefix);
+            }
+        }
+
+        for nullifier in nullifiers {
+            let prefix = self.nullifier_idx.remove(&nullifier).unwrap();
+            if self.accounts.get_mut(&prefix).unwrap().commit_nullifier(nullifier).is_empty() {
+                self.remove_account(prefix);
+            }
+        }
+    }
+
+    /// Handles [`MempoolEvent::TransactionsReverted`] events.
+    fn revert_transaction(&mut self, tx: TransactionId) {
+        let Impact { account_delta, notes, nullifiers } =
+            self.inflight_txs.remove(&tx).unwrap_or_default();
```

Comment:
Would the `or_default()` part be unreachable here? Basically, if we are reverting a transaction, isn't it guaranteed that is is in `self.inflight_txs`?

Reply:
The mempool sends _all_ reverted transactions but we only store txs with network-related impact. A tx with no network account delta and which produces no network notes or network nullifiers would not be in `self.inflight_txs`.

Might be clearer if I rewrite this as an early return with a comment though.

```rust
// Ignore transactions that aren't tracked.
//
// We only track transactions that have an impact on network state.
let Some(Impact { account_delta, notes, nullifier }) = self.inflight_txs.remove(&tx) else {
    return;
};
```

Reply:
Makes sense! Yes, using early return would make things more clear here.